insights/v1alpha2/types_insights.go

```diff
@@ -0,0 +1,446 @@
+// +kubebuilder:validation:Optional
+// +kubebuilder:validation:Enum=Running;Completed;Failed;Pending
+// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Running' && self == 'Pending')", message="dataGatherState cannot transition from Running to Pending"
+// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Completed' && self == 'Pending')", message="dataGatherState cannot transition from Completed to Pending"
+// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Failed' && self == 'Pending')", message="dataGatherState cannot transition from Failed to Pending"
+// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Completed' && self == 'Running')", message="dataGatherState cannot transition from Completed to Running"
+// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Failed' && self == 'Running')", message="dataGatherState cannot transition from Failed to Running"
+type DataGatherState string
+
+const (
+	// Data gathering is running
+	DataGatherStateRunning DataGatherState = "Running"
+	// Data gathering is completed
+	DataGatherStateCompleted DataGatherState = "Completed"
+	// Data gathering failed
+	DataGatherStateFailed DataGatherState = "Failed"
+	// Data gathering is pending
+	DataGatherStatePending DataGatherState = "Pending"
+)
+
+// DataGatherStatus contains information relating to the DataGather state.
+// +kubebuilder:validation:XValidation:rule="(!has(oldSelf.insightsRequestID) || has(self.insightsRequestID))",message="cannot remove insightsRequestID attribute from status"
+// +kubebuilder:validation:XValidation:rule="(!has(oldSelf.startTime) || has(self.startTime))",message="cannot remove startTime attribute from status"
+// +kubebuilder:validation:XValidation:rule="(!has(oldSelf.finishTime) || has(self.finishTime))",message="cannot remove finishTime attribute from status"
+// +kubebuilder:validation:XValidation:rule="(!has(oldSelf.dataGatherState) || has(self.dataGatherState))",message="cannot remove dataGatherState attribute from status"
+// +kubebuilder:validation:Optional
+type DataGatherStatus struct {
+	// conditions is an optional field that provides details on the status of the gatherer job.
+	// It may not exceed 100 items and must not contain duplicates.
+	//
+	// The current condition types are DataUploaded, DataRecorded, DataProcessed, RemoteConfigurationNotAvailable, RemoteConfigurationInvalid
+	//
+	// The DataUploaded condition is used to represent whether or not the archive was successfully uploaded for further processing.
+	// When it has a status of True and a reason of HttpStatus200, the archive was successfully uploaded.
+	// When it has a status of Unknown and a reason of NoUploadYet, the upload has not occurred yet.
+	// When it has a status of False and a reason like HttpStatusXXX, the upload failed and the reason reflects the returned HTTP status code. The accompanying message will include the specific error encountered.
```

Comment:
What about a reason like `UploadFailure` instead of having a dynamic reason that is based on the HTTP status code you receive? 

Having something dynamic like that makes it more difficult for someone to write automation around this status condition because they have to encode all the possible `HttpStatusXXX` values into their automation as opposed to a single static value that represents roughly the same thing (i.e this thing failed and that is bad).

Reply:
@everettraven what would be your suggestion on how to propagate the exact HTTP status code from the dataGather resource please? It can go to the message attribute value, but I've always thought the reason is the attribute for "machines/programs" to read.

Reply:
Why do users/automation need to know the exact HTTP status codes?

Reply:
FWIW, I have a similar understanding of the fields.